tofu/geom/_GG.pyx

```diff
@@ -4932,3 +4943,481 @@ def compute_solid_angle_map(double[:,::1] part_coords, double[::1] part_r,
+    cdef long* ncells_rphi = NULL
+    cdef double* disc_r0 = NULL
+    cdef double* disc_r  = NULL
+    cdef double* disc_z  = NULL
+    cdef double* step_rphi = NULL
+    cdef double** data = NULL
+    cdef np.ndarray[long, ndim=1] ind
+    cdef np.ndarray[double, ndim=1] reso_rdrdz
+    cdef np.ndarray[double, ndim=1] dot_GBGC
+    cdef np.ndarray[double, ndim=2] pts
+    cdef np.ndarray[double, ndim=2] sa_map
+    cdef np.ndarray[double, ndim=2] vec_GB
+    cdef np.ndarray[double, ndim=2] vec_GC
+    cdef np.ndarray[double, ndim=2] centroids
+    cdef np.ndarray[double, ndim=2] poly_lnorms_tot
+    cdef np.ndarray[double, ndim=2] cross_GBGC
+    cdef double[:, ::1] temp
+    print("begining.....................................")
+    #
+    # == Testing inputs ========================================================
+    if test:
+        if block:
+            msg = "ves_poly and ves_norm are not optional arguments"
+            assert ves_poly is not None and ves_norm is not None, msg
+            bool1 = (ves_poly.shape[0]==2 and ves_norm.shape[0]==2
+                     and ves_norm.shape[1]==ves_poly.shape[1]-1)
+            msg = "Args ves_poly & ves_norm must be of the same shape (2, NS)!"
+            assert bool1, msg
+            bool1 = (lstruct_lims is None
+                     or len(lstruct_normy) == len(lstruct_normx))
+            bool2 = (lstruct_normx is None
+                     or len(lstruct_polyx) == len(lstruct_polyy))
+            msg = "Args lstruct_polyx, lstruct_polyy, lstruct_lims,"\
+                + " lstruct_normx, lstruct_normy, must be None or"\
+                + " lists of same len()!"
+            assert bool1 and bool2, msg
+        msg = "[eps_uz,eps_vz,eps_a,eps_b] must be floats < 1.e-4!"
+        assert all([ee < 1.e-4 for ee in [eps_uz, eps_a,
+                                          eps_vz, eps_b,
+                                          eps_plane]]), msg
+        msg = "ves_type must be a str in ['Tor','Lin']!"
+        assert ves_type.lower() in ['tor', 'lin'], msg
+    # ...
+    print("after tests.................................")
+    # .. Formatting polygon coordinates ........................................
+    npoly = len(poly_coords)
+    for ii in range(npoly):
+        print(poly_coords[ii])
+        poly_coords[ii] = format_poly(poly_coords[ii], Clock=False,
+                                      close=False, Test=True)
+        print(poly_coords[ii])
+    print("......................... after format")
+    # .. Dividing polys in triangles ...........................................
+    ltri = <long**>malloc(sizeof(long*) * npoly)
+    # re writting_polygons coordinates to C type:
+    data = <double **> malloc(npoly * sizeof(double *))
+    for ii in range(npoly):
+        temp = poly_coords[ii]
+        data[ii] = &temp[0, 0]
+    _vt.triangulate_polys(
+        &data[0],
+        &poly_lnvert[0],
+        npoly,
+        ltri,
+        num_threads
+    )
```

Comment:
@lasofivec, here the routine  _vt.triangulate_polys() is called with the first argument as `&data[0]`
Here data is defined as `data = <double **> malloc(npoly * sizeof(double *))`

But in line 2812 (in routine `def vignetting()`, the same routine is called with the first argument as `data`.
And there data is defined as `data = <double **> malloc(nvign*sizeof(double *))`

So in both cases, data refers a similar object.
But in here it is provided as a memory address of the first element (`&data[0]`), while in line 2812 it is provided directly (`data`)

Questions:
* Why this difference ?
* Is it a bug ? 
* Which one is correct ?

The definition of the routine, in `_vignetting_tools.pyx `(line 291) says it should have a double pointer of type `double**` as first argument. Which leads to think that only the first case (here, line 5170 is correct).

More questions:
* Why use a double pointer instead of a single pointer in the routine definition ?
* Why provide a pointer only to the first element to data ?